rfcs/text/1020-link-entities.md

```diff
@@ -0,0 +1,385 @@
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+In brief, this RFC removes the current definition of _link types_, and extends the definition of an _entity type_, and of an _entity_, as follows:
+
+## Entity Types
+
+An ontology type (element of the Type System), represented as a JSON schema, that defines the structure of an entity, its link entities and their endpoints.
+
+### Link Entity Types
+
+A _link entity type_ is an entity type which includes the following field in its schema:
+
+```json5
+"allOf": [
+  {
+    "ref": "https://blockprotocol.org/@blockprotocol/types/entity-type/link/v/1"
+  }
+]
+```
+
+The reasoning for this representation is explored in slightly more detail in the [Reference-level explanation](#reference-level-explanation) section below, but at a high-level this is just a marker annotation at the moment, although it will hopefully be future compatible with [type inheritance](https://github.com/blockprotocol/blockprotocol/pull/428).
+
+## Entity
+
+As previously defined in the spec, an _entity_ is a combination of **metadata** and **properties**.
+An example of one is:
+
+```json5
+{
+  properties: {
+    "https://example.com/@example/types/property-type/shortname/": "Bob",
+  },
+  metadata: {
+    recordId: {
+      entityId: "c6f0f771-b30b-4557-8979-e339006ac793",
+      editionId: "8add2c02-6295-4c6b-b079-849f3c403ad7",
+    },
+    entityTypeId: "https://example.com/@example/types/entity-type/user/v/1",
+  },
+}
+```
+
+### Link Entity
+
+A _link entity_ is a special kind of entity that represents a link between two other entities.
+It contains an additional field (alongside `"properties"` and `"metadata'`) `"linkData"` which looks like the following:
+
+```json5
+"linkData": {
+  "leftEntityId": "a28041f9-f8c5-4275-a601-ac5cb0d57a0d",
+  "rightEntityId": "6d7fdb6c-f519-4554-a6d7-485c818c7a63",
+  "leftToRightOrder": 2
+  "rightToLeftOrder": 10
+},
+```
+
+The fields within `"linkData"` are as follows:
+
+- `"leftEntityId"`: The `entityId` of the entity on the "left" hand side of the link entity. In most scenarios this will be the "source" of the link entity.
+- `"rightEntityId"`: The `entityId` of the entity on the "right" hand side of the link entity. In most scenarios this will be the "target" of the link entity.
+- `"leftToRightOrder"`: A value used to order this link entity against other link entities of the same type with the same `leftEntityId`.
+- `"rightToLeftOrder"`: A value used to order this link entity against other link entities of the same type with the same `rightEntityId`.
+
+The choice of "left" and "right" as opposed to "source" and "target" is an optimistic attempt to leave room for an extension of the system in the future to allow for "bi-directional" links (where a "source" and "target" would not make sense).
```

Comment:
Changing to a hypergraph (where an edge can connect to multiple vertices on both ends) would be a pretty substantial change, and not one I've heard a great argument for at the moment. This seems better modelled (to me) with high-order-abstractions such as a `Group`, etc. as that is more extensible.

However, I don't see how allowing multiple endpoints would actually support bidirectional links, if you have

```
Sources   Link         Destinations
[Bob, ] --Friend Of--> [Charlie, ] 
```

I don't see how that gives us the bidirectionality, were you imagining us then doing something like this?
```
Sources          Link         Destinations
[Bob, Charlie] --Friend Of--> [Charlie, Bob] 
```

Because that would seem to imply `Bob` is friends with `Bob`, and if we established explicit mappings within the lists (adding a _**lot**_ of complexity), then we're basically back to just having an ordered list of links (`[Bob --Friend of--> Charlie, Charlie --Friend of--> Bob]`)

Reply:
I was imagining that, and that you would simply exclude/ignore self-referential links. Is there an instance an entity would ever actually want to link to itself?